storage_service: refresh_sync_nodes: restrict to normal token owners

bhalevy opened the pull request:
It is possible that topology will contain nodes that are no longer normal token owners, so they don't need to be sync'ed with.

Fixes scylladb/scylladb#14793

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2631/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2632/

asias commented on locator/token_metadata.cc:
```diff
@@ -556,6 +556,9 @@ std::unordered_map<inet_address, host_id> token_metadata_impl::get_endpoint_to_h
     std::unordered_map<inet_address, host_id> map;
     map.reserve(nodes.size());
     for (const auto& [endpoint, node] : nodes) {
+        if (!is_normal_token_owner(endpoint)) {
```
Why do you want to restrict to normal token owners?

This function is only used by

```
      {     
         "path":"/storage_service/host_id",
         "operations":[
            {
               "method":"GET",
               "summary":"Retrieve the mapping of endpoint to host ID",
               "type":"array",
               "items":{
                  "type":"mapper"
               },
               "nickname":"get_host_id_map",
               "produces":[
                  "application/json"
               ],    
               "parameters":[
               ]
            }
         ]        
      },
```

Which returns a map of ip <> host_id?

This api does not say it returns only the node that is part of the ring?

For example, when a node is bootstrapping and not part of the ring, it should be returned from the list. 

After this change, the api does not return the node at all.

bhalevy replied:
Because it caused a raft test to fail since it returned a decommissioned node with null host_id.

asias replied:
But this is not a reason to change an API behavior silently.

bhalevy replied:
Good point, will fix.
But I will still restrict the output to nodes with known host_id since the api is about "Retrieve the mapping of endpoint to host ID"

asias commented on service/storage_service.cc:
```diff
@@ -3551,7 +3551,7 @@ class node_ops_ctl {
             // FIXME: use node* rather than endpoint
             auto node = np->endpoint();
-            if (!ignore_nodes.contains(node) && sync_to_node(node)) {
+            if (!ignore_nodes.contains(node) && tmptr->is_normal_token_owner(node) && sync_to_node(node)) {
```
What nodes are stored in get_topology()?   What's the point of using get_topology::for_each_node() and filter with tmptr->is_normal_token_owner? We already have tmptr->get_all_endpoints() to return all normal token owners nodes?




```
    // Returns nodes that are officially part of the ring. It does not include
    // node that is still joining the cluster, e.g., a node that is still
    // streaming data before it finishes the bootstrap process and turns into
    // NORMAL status.
    const std::unordered_set<inet_address>& get_all_endpoints() const noexcept {
        return _normal_token_owners;
    }
```

bhalevy replied:
The topology may hold pending and leaving nodes as well as left nodes.
The reason to use topology rather than `get_all_endpoints()` is the transition to use topology::node that has both the ip address and host_id (as well as dc/rack).
The end game indeed is planned be to use nodes/host_ids in token_metadata so `get_all_endpoints` or a similar function would be able to return a set of nodes instead of endpoints, but we're not there yet.
But it might be a good idea to provide this interface now, e.g. `get_all_nodes()` that can construct a unordered_set of nodes which are normal token owners.

asias replied:
Are you talking about making tm::get_all_nodes return a set of locator::node?

BTW, we even have locator::node::_state. It makes less sense to query it again with tmptr->is_normal_token_owner(node) instead of checking _state directly. Is _state accurate enough about the state? Why not check the flag directly?

```
enum class state {                                             
    none = 0,                                                  
    joining,    // while bootstrapping, replacing              
    normal,                                                    
    leaving,    // while decommissioned, removed, replaced     
    left        // after decommissioned, removed, replaced     
};                  
```

bhalevy replied:
The node state should be correct, but it `leaving` doesn't distinguish between decommissioned nodes to removed nodes.   Maybe it's a good opportunity to make this distinction if possible and here we'd need to return all nodes in `normal` or `decommissioning` state.

bhalevy replied:
> Are you talking about making tm::get_all_nodes return a set of locator::node?

Yes.

bhalevy commented:
In v2 ([b59bb53](https://github.com/scylladb/scylladb/commit/b59bb534b074374831185f8d6e73dcabf6c4bbc6))
- make node::state fine grained so we can distinguish between different `leaving` states
- refresh_sync_nodes: restrict nodes to `is_token_owner_for_read` (a predicate introduced in the previous patch)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2671/

asias commented on service/storage_service.cc:
```diff
@@ -1875,7 +1875,7 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
                 get_broadcast_address() == *replace_address ? "the same" : "a different",
                 get_broadcast_address(), *replace_address);
-            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::leaving);
+            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::being_replaced);
```
If a node replaces using the same ip address of the node being replaced, should we call it being_replaced or replacing?

asias replied:
If we mark it as being_replaced,  is_token_owner_for_read() is going to return true for the node being replaced which is supposed to be dead and we should not read from it from the perspective node ops. 

```
    bool is_token_owner_for_read() const noexcept {
        switch (_state) {
        case state::normal:
        case state::being_decommissioned:
        case state::being_replaced:
            return true;
        default:
            return false;
        }
    }
```

bhalevy replied:
> If a node replaces using the same ip address of the node being replaced, should we call it being_replaced or replacing?

Now that we don't reuse the replaced node host id, there will be 2 node objects in topology, distinguished by their host_id.
The one which state is updated here is always the one being replaced.

bhalevy replied:
> If we mark it as being_replaced, is_token_owner_for_read() is going to return true for the node being replaced which is supposed to be dead and we should not read from it from the perspective node ops.

You're 100% right. It's a bug and it shouldn't be returned there.

bhalevy replied:
fixed in v3

asias replied:
On one hand, removing being_replaced from is_token_owner() will exclude the node from nodes to sync in replace. On the other hand, node being_replaced is still the owner of the tokens, it is wrong not to include it. 

I think it is better we keep the token ownership correct and filter the node being replaced with the filter. 

Actually we already have such filter

    // Step 1: Decide who needs to sync data for replace operation
    // The replacing node is not a normal token owner yet
    // Add it back explicitly after checking all other nodes.
    ctl.start("replace", [&] (gms::inet_address node) {
        return node != replace_address;
    });

bhalevy replied:
But isn't the replaced node required to be dead to be replaced?  If so, it shouldn't be sync'ed with.
Also, if we include replaced nodes we should include also removed nodes since they are officially still token owners until removed / replaced.  The point of `is_token_owner_for_read` was to restrict to nodes that should be accessed for read - and replaced and removed nodes are not part of that set.

bhalevy replied:
Note: replaced may be in the future, with tablets, if we allow replacing live nodes.
But since bootstrapping a new node with tablets is quick and adds the new node as empty - and only later uses the tablets load balancer to fill it, it will be feasible to add one or more new nodes and then simply decommission the node we want to replace.  That would be a close enough approximation of replacing a live node.
/Cc @tgrabiec

asias commented on locator/topology.hh:
```diff
@@ -102,6 +105,59 @@ public:
+
+    bool is_token_owner_for_read() const noexcept {
+        switch (_state) {
+        case state::normal:
```
Currently, the rule is simple, if a node is in NORMAL status it is part of the ring. Here we reuse the term normal for a state that is a bit different.  Nothing wrong about it. But it turns out, even with the much more complicated states, is_token_owner_for_read() and is_token_owner_for_write() are not very useful, as discussed below.

asias replied:
How about just leave `bool is_token_owner()` and drop the _for_read and _for_write versions?

bhalevy replied:
That"s possible.  It'd essentially be the "for read" version, while the "for write" version is a super-set of it that includes also pending nodes.

bhalevy replied:
done in v3

bhalevy commented:
In v3 ([73ae49c](https://github.com/scylladb/scylladb/commit/73ae49cbed0630a9bc1c73fba167833565f9e833)):
- locator: topology: node::state: make fine grained
  - dropped `is_token_owner_for_{read,write}()`
  - added simplified `is_token_owner()` that is fixed to not contain `being_replaced`
- storage_service: refresh_sync_nodes: fix log message
- storage_service: refresh_sync_nodes: restrict to token owners for read
  - use `node.is_token_owner()`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2689/

bhalevy commented:
@asias, @tgrabiec please re-review

asias commented on service/storage_service.cc:
```diff
@@ -3563,7 +3563,7 @@ class node_ops_ctl {
             // FIXME: use node* rather than endpoint
             auto node = np->endpoint();
-            if (!ignore_nodes.contains(node) && sync_to_node(node)) {
+            if (!ignore_nodes.contains(node) && np->is_token_owner() && sync_to_node(node)) {
```
The commit message still says is_token_owner_for_read

bhalevy replied:
fixed that (and the changelog message)

asias commented:
On Mon, Jul 31, 2023 at 2:50 PM Benny Halevy ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In service/storage_service.cc
> <https://github.com/scylladb/scylladb/pull/14798#discussion_r1278857177>:
>
> > @@ -1875,7 +1875,7 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
>              slogger.info("Replacing a node with {} IP address, my address={}, node being replaced={}",
>                  get_broadcast_address() == *replace_address ? "the same" : "a different",
>                  get_broadcast_address(), *replace_address);
> -            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::leaving);
> +            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::being_replaced);
>
> But isn't the replaced node required to be dead to be replaced? If so, it
> shouldn't be sync'ed with.
> Also, if we include replaced nodes we should include also removed nodes
> since they are officially still token owners until removed / replaced. The
> point of is_token_owner_for_read was to restrict to nodes that should be
> accessed for read - and replaced and removed nodes are not part of that set.
>

I am not sure if you saw the following comments. It shows pending on the GH
UI:

All the leaving states
```
   state::being_decommissioned:
   state::being_removed:
   state::being_replaced:
```
and the normal states

```
  state::normal
```
are still token owners, no?


The users of the is_token_owner() is not limited to how we calculate who to
sync with. We should not change the generic is_token_owner() to suit sync
nodes only.


-- 
Asias

bhalevy commented:
> The users of the is_token_owner() is not limited to how we calculate who to
sync with. We should not change the generic is_token_owner() to suit sync
nodes only.

Fair enough.  We can remove this interface for now as it has no other users and
just implement the predicate in the node_ops_ctl layer

bhalevy commented:
or give it another name. If is_token_owner_for_read is unclear, how about `is_live_token_owner`?

asias commented:
> or give it another name. If `is_token_owner_for_read` is unclear, how about `is_live_token_owner`?

This name is confusing too, e.g, a normal node is down but we still should include it in nodes to sync with.

asias commented:
> > The users of the is_token_owner() is not limited to how we calculate who to
> > sync with. We should not change the generic is_token_owner() to suit sync
> > nodes only.
> 
> Fair enough. We can remove this interface for now as it has no other users and just implement the predicate in the node_ops_ctl layer

OK. Yes, I think it is best to handle at the node_ops_ctrl layer. It is very obvious and easy to understand at this layer, e.g., we should not sync with the node being replaced.

asias commented:
BTW, I think we just hit a GH bug. I was commenting while you updated a new version. I get this pending comments forever. 

![image](https://github.com/scylladb/scylladb/assets/320076/a4644507-f5fc-439c-87c0-752700133565)

bhalevy commented on locator/topology.hh:
```diff
@@ -102,6 +105,45 @@ public:
+        }
+    }
+
+    bool is_token_owner() const noexcept {
```
Now I can see this comment.

asias replied:
ok.

asias commented on service/storage_service.cc:
```diff
@@ -3770,11 +3770,23 @@ class node_ops_ctl {
+        auto can_sync_with_node = [] (const locator::node& node) {
+            // Sync with reachable token owners.
+            // Note that although nodes in `being_replaced` and `being_removed`
+            // are still token owners, they are known to be dead and can't be sync'ed with.
```
Good! This is way clearer!

asias reviewed (approved):
Looks good to me. Thanks!

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2770/

bhalevy commented:
@scylladb/scylla-maint please merge
